src/libostree/ostree-repo-pull.c

```diff
@@ -478,6 +488,33 @@ scan_object_queue_data_free (ScanObjectQueueData *scan_data)
   g_free (scan_data);
 }
 
+/* Check whether a particular operation should be retried. This is entirely
+ * based on how it failed (if at all) last time, and whether the operation has
+ * some retries left. The retry count is set when the operation is first
+ * created, and must be decremented by the caller. (@n_retries_remaining == 0)
+ * will always return %FALSE from this function. */
+static gboolean
+should_retry_request (const GError *error,
+                      guint         n_retries_remaining)
+{
+  if (error == NULL || n_retries_remaining == 0)
+    return FALSE;
+
+  /* Return TRUE for transient errors. */
+  if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_TIMED_OUT) ||
+      g_error_matches (error, G_IO_ERROR, G_IO_ERROR_HOST_NOT_FOUND) ||
+      g_error_matches (error, G_IO_ERROR, G_IO_ERROR_HOST_UNREACHABLE) ||
+      g_error_matches (error, G_RESOLVER_ERROR, G_RESOLVER_ERROR_NOT_FOUND) ||
+      g_error_matches (error, G_RESOLVER_ERROR, G_RESOLVER_ERROR_TEMPORARY_FAILURE))
```

Comment:
Wait, I'm confused now how this actually works. I don't see anywhere in either the libsoup or the libcurl fetchers where we translate curl/soup specific transport errors into these. We need to expand handling in those functions to tease out these transient issues, right? E.g. in the libcurl fetcher, I would expect to see something like:

```c
      else if (curlres != CURLE_OK)
        {
           switch (curlres)
             {
             case CURLE_COULDNT_RESOLVE_HOST:
                /* map to G_RESOLVER_ERROR_NOT_FOUND */
                break;
             case CURLE_COULDNT_CONNECT:
                /* map to G_IO_ERROR_HOST_UNREACHABLE */
                break;
             }
        }
```

or something similar...

Reply:
In the libsoup case, I guess I’m getting the `GSocket` errors propagated back up through libsoup. I could definitely improve `on_request_sent()` in `ostree-fetcher-soup.c` to propagate some of the higher level errors a bit more successfully though.

I haven’t tested with the libcurl backend.

Reply:
I'm not sure it's worth trying to maintain a full mapping between the error domains, could just squash the ones we want to retry from curl to `G_IO_ERROR_HOST_UNREACHABLE` or whatever.